src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -282,166 +292,195 @@ struct TreeVisitor<'tree> {
+                self.stack.pop();
+                self.propagate_at(this, idx, rel_pos)?;
+            } else {
+                *is_final = true;
+                let handle_children = self.should_continue_at(this, idx, rel_pos);
+                match handle_children {
+                    ContinueTraversal::Recurse => {
+                        // add all children, and also leave the node itself
+                        // on the stack so that it can be visited later.
+                        let node = this.nodes.get(idx).unwrap();
+                        for &child in node.children.iter() {
+                            self.stack.push((child, rel_pos, false));
+                        }
+                    }
+                    ContinueTraversal::SkipSelfAndChildren => {
+                        // skip self
+                        self.stack.pop();
+                        continue;
+                    }
+                }
+            }
         }
         Ok(())
     }
 
-    /// Push all ancestors to the exploration stack in order of nearest ancestor
-    /// towards the top.
-    fn push_and_visit_strict_ancestors(
-        &mut self,
-        visitor: &mut TreeVisitor<'_>,
-    ) -> Result<(), OutErr> {
-        let mut path_ascend = Vec::new();
-        // First climb to the root while recording the path
-        let mut curr = self.initial;
-        while let Some(ancestor) = visitor.nodes.get(curr).unwrap().parent {
-            path_ascend.push((ancestor, curr));
-            curr = ancestor;
-        }
-        // Then descend:
-        // - execute f_propagate on each node
-        // - record children in visit
-        while let Some((ancestor, next_in_path)) = path_ascend.pop() {
-            // Explore ancestors in descending order.
-            // `next_in_path` is excluded from the recursion because it
-            // will be the `ancestor` of the next iteration.
-            // It also needs a different `AccessRelatedness` than the other
-            // children of `ancestor`.
-            self.exec_and_visit(
-                visitor,
-                ancestor,
-                Some(next_in_path),
-                AccessRelatedness::StrictChildAccess,
-            )?;
-        }
-        Ok(())
+    fn new(
+        initial: UniIndex,
+        f_continue: NodeContinue,
+        f_propagate: NodeApp,
+        err_builder: ErrHandler,
+    ) -> Self {
+        Self { initial, f_continue, f_propagate, err_builder, stack: Vec::new() }
     }
 }
 
 impl<'tree> TreeVisitor<'tree> {
-    // Applies `f_propagate` to every vertex of the tree top-down in the following order: first
-    // all ancestors of `start`, then `start` itself, then children of `start`, then the rest.
+    // Applies `f_propagate` to every vertex of the tree bottom-up in the following order: first
+    // all ancestors of `start` (starting with `start` itself), then children of `start`, then the rest,
+    // always going bottom-up.
```

Comment:
I am confused. It may be good to have a little image here or so explaining the order. It seems like we're jumping up and down the tree pretty wild here? First move up from `start`, then jump to the bottom and move towards `start`, and jump to other leaves and move towards ancestors of `start`?

Maybe call this "piecewise bottom-up" to make it clear that this is not a normal bottom-up traversal.

That's a very strange iteration order. I guess the key point is that for TB correctness, the order does not matter, so this only affects error messages? If so, then please say that explicitly.

Reply:
I added an ASCII art iamge.